File: go/2784.go
package main

func numBeeg(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func isGood(nums []int) bool {
	maxi := nums[0]

	for i := 1; i < len(nums); i++ {
		maxi = numBeeg(nums[i], maxi)
	}

	if len(nums) < maxi+1 {
		return false
	}

	freq := make([]int, maxi+1)
	for i := 0; i < len(nums); i++ {
		freq[nums[i]]++
	}

	for i := 1; i < maxi; i++ {
		if freq[i] != 1 {
			return false
		}
	}

	return freq[maxi] == 2
}

File: zig/2784.zig
const std = @import("std");

fn solution(allocator: std.mem.Allocator, nums: []const i32) !bool {
    var maxi = nums[0];

    for (1..nums.len) |i| {
        maxi = @max(maxi, nums[i]);
    }

    if (nums.len < maxi + 1) {
        return false;
    }

    const maxu = @as(usize, @intCast(maxi));

    var freq = try allocator.alloc(i32, maxu + 1);
    @memset(freq, 0);

    defer allocator.free(freq);

    for (nums) |num| {
        freq[@as(usize, @intCast(num))] += 1;
    }

    for (1..maxu) |i| {
        if (freq[i] != 1) {
            return false;
        }
    }

    return freq[maxu] == 2;
}

const expect = std.testing.expectEqual;

test "test case #1" {
    const allocator = std.testing.allocator;

    const nums = [_]i32{ 2, 1, 3 };
    const expected = false;
    const actual = try solution(allocator, &nums);

    try expect(expected, actual);
}

test "test case #2" {
    const allocator = std.testing.allocator;

    const nums = [_]i32{ 1, 3, 3, 2 };
    const expected = true;
    const actual = try solution(allocator, &nums);

    try expect(expected, actual);
}

test "test case #3" {
    const allocator = std.testing.allocator;

    const nums = [_]i32{ 3, 4, 4, 1, 2, 1 };
    const expected = false;
    const actual = try solution(allocator, &nums);

    try expect(expected, actual);
}
